File: 프로그래머스/3/150366. 표 병합/표 병합.java
import java.util.*;

class Solution {
    static final int SIZE = 51; // 1-based 인덱스, 50까지

    public String[] solution(String[] commands) {
        int[][] merged = new int[SIZE][SIZE]; // 병합 그룹 대표 좌표를 1차원 인덱스로 저장
        String[][] content = new String[SIZE][SIZE];

        // 초기화: 각 셀이 자기 자신이 대표, 내용은 EMPTY
        for (int i = 1; i < SIZE; i++) {
            for (int j = 1; j < SIZE; j++) {
                merged[i][j] = i * SIZE + j;
                content[i][j] = "EMPTY";
            }
        }

        List<String> answer = new ArrayList<>();

        for (String cmd : commands) {
            String[] parts = cmd.split(" ");
            if (parts[0].equals("UPDATE")) {
                if (parts.length == 4) {
                    int r = Integer.parseInt(parts[1]);
                    int c = Integer.parseInt(parts[2]);
                    String value = parts[3];
                    int group = merged[r][c];
                    // 병합 그룹 전체 값을 바꿔준다
                    for (int i = 1; i < SIZE; i++) {
                        for (int j = 1; j < SIZE; j++) {
                            if (merged[i][j] == group) {
                                content[i][j] = value;
                            }
                        }
                    }
                } else { // UPDATE value1 value2
                    String value1 = parts[1];
                    String value2 = parts[2];
                    for (int i = 1; i < SIZE; i++) {
                        for (int j = 1; j < SIZE; j++) {
                            if (content[i][j].equals(value1)) {
                                content[i][j] = value2;
                            }
                        }
                    }
                }
            } else if (parts[0].equals("MERGE")) {
                int r1 = Integer.parseInt(parts[1]);
                int c1 = Integer.parseInt(parts[2]);
                int r2 = Integer.parseInt(parts[3]);
                int c2 = Integer.parseInt(parts[4]);
                if (r1 == r2 && c1 == c2) continue; // 자기 자신과 병합은 무시

                int group1 = merged[r1][c1];
                int group2 = merged[r2][c2];

                // 병합 시 대표 선택(둘 중 한 그룹으로 통합)
                String newValue = !content[r1][c1].equals("EMPTY") ? content[r1][c1] : content[r2][c2];

                for (int i = 1; i < SIZE; i++) {
                    for (int j = 1; j < SIZE; j++) {
                        if (merged[i][j] == group2) {
                            merged[i][j] = group1;
                        }
                    }
                }
                // 병합된 그룹 전체 값을 newValue로 변경
                for (int i = 1; i < SIZE; i++) {
                    for (int j = 1; j < SIZE; j++) {
                        if (merged[i][j] == group1) {
                            content[i][j] = newValue;
                        }
                    }
                }
            } else if (parts[0].equals("UNMERGE")) {
                int r = Integer.parseInt(parts[1]);
                int c = Integer.parseInt(parts[2]);
                int group = merged[r][c];
                String cellValue = content[r][c];
                // 병합된 모든 셀을 분리
                for (int i = 1; i < SIZE; i++) {
                    for (int j = 1; j < SIZE; j++) {
                        if (merged[i][j] == group) {
                            merged[i][j] = i * SIZE + j;
                            content[i][j] = "EMPTY";
                        }
                    }
                }
                // 단, 해제된 셀(r,c)는 원래 값으로 복원
                content[r][c] = cellValue;
            } else if (parts[0].equals("PRINT")) {
                int r = Integer.parseInt(parts[1]);
                int c = Integer.parseInt(parts[2]);
                answer.add(content[r][c]);
            }
        }

        return answer.toArray(new String[0]);
    }
}

File: 프로그래머스/3/150366. 표 병합/표 병합.py
table = [["EMPTY"]*51 for _ in range(51)]
par = [[[r, c] for c in range(51)] for r in range(51)]
output = []

def find(r, c):
    if par[r][c] == [r, c]:
        return [r, c]
    pr, pc = par[r][c]
    par[r][c] = find(pr, pc)
    return par[r][c]


def MERGE(r1, c1, r2, c2):
    if (r1, c1) == (r2, c2):
        return
    r1, c1 = find(r1, c1)
    r2, c2 = find(r2, c2)
    if (r1, c1) == (r2, c2):
        return

    if table[r1][c1] == "EMPTY" and table[r2][c2] != "EMPTY":
        r1, c1, r2, c2 = r2, c2, r1, c1

    group = []
    for i in range(1, 51):
        for j in range(1, 51):
            pi, pj = find(i, j)
            if (pi, pj) == (r2, c2):
                group.append((i, j))
    for i, j in group:
        par[i][j] = [r1, c1]


def UNMERGE(r, c):
    pr, pc = find(r, c)
    v = table[pr][pc]
    
    group = []
    for i in range(1, 51):
        for j in range(1, 51):
            pi, pj = find(i, j)
            if [pi, pj] == [pr, pc]:
                group.append((i, j))
    for i, j in group:
        par[i][j] = [i, j]
        table[i][j] = "EMPTY"
    table[r][c] = v


def UPDATE(r, c, v):
    pr, pc = find(r, c)
    table[pr][pc] = v


def UPDATE_VAL(val1, val2):
    for r in range(1, 51):
        for c in range(1, 51):
            if table[r][c] == val1:
                table[r][c] = val2


def PRINT(r, c):
    pr, pc = find(r, c)
    output.append(table[pr][pc])


def solution(commands):
    for command in commands:
        op, *args = command.split()
        if (op == "UPDATE"):
            if len(args) == 3:
                r, c = map(int, args[:2])
                UPDATE(r, c, args[2])
            else:
                UPDATE_VAL(*args)
        if (op == "MERGE"):
            MERGE(*map(int, args))
        if (op == "UNMERGE"):
            UNMERGE(*map(int, args))
        if (op == "PRINT"):
            PRINT(*map(int, args))

    return output